### DISCUSSION TITLE
OpenResty 无法启动

### DISCUSSION THREAD

**Discussion by purainity** [1Panel-dev/1Panel#5782] [Q&A] [answered]
### 1Panel 版本

社区版：v1.10.11-lts

### 问题描述

OpenResty 无法启动，报错

### 重现步骤

前往应用商店安装 OpenResty，安装成功后点击启动

### 相关日志输出

```shell
服务内部错误: stderr: Container 1Panel-openresty-Lb0D Starting Error response from daemon: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting "/ext_overlay/Configs/1Panel/1panel/apps/openresty/openresty/conf/nginx.conf" to rootfs at "/usr/local/openresty/nginx/conf/nginx.conf": mount /ext_overlay/Configs/1Panel/1panel/apps/openresty/openresty/conf/nginx.conf:/usr/local/openresty/nginx/conf/nginx.conf (via /proc/self/fd/6), flags: 0x5000: not a directory: unknown: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type
```

### 附加信息

截图

![Screenshot_2024-07-11-14-40-26-907_com android browser](https://github.com/1Panel-dev/1Panel/assets/126877770/cb3d57f2-3d24-45f5-80c1-8100ef1a15ee)

**Comment by zhengkunwang223** [comment]
什么操作系统和架构？什么机器？

**Comment by purainity** [comment]
> 什么操作系统和架构？什么机器？ 

x86_64 的架构，是 iStoreOS 中通过 Docker 的方式安装的 1Panel

**Comment by wanghe-fit2cloud** [comment]
> error mounting "/ext_overlay/Configs/1Panel/1panel/apps/openresty/openresty/conf/nginx.conf" to rootfs

可以先尝试手动创建应用的安装目录。

**Reply by purainity** [reply]
看了下，这个安装目录是有正常的配置文件的

**Comment by nickdlkk** [comment]
应该是IstoreOS安装时映射的问题，iStoreOS启动1Panel时设置的Docker映射里有`/root/Configs/1Panel:/iStorePane`，前面的目录是安装时配置的，后面的目录是固定的，到容器内的`/iStorePane`

> 见 https://github.com/linkease/openwrt-app-actions/blob/24f01cb47525e2e8c0677ef0ec8563a1c8487348/applications/luci-app-istorepanel/root/usr/libexec/istorec/istorepanel.sh#L55

所以1Panel在内部创建的OpenResty目录`/root/Configs/1Panel/1panel/apps/openresty/openresty` 并没有映射到宿主机上(iStoreOS)

而1Panel使用Docker创建OpenResty容器的时候，使用的是宿主机的目录，所以找不到文件


解决办法：

IstoreOS里面编辑1panel的那个容器，新增一个`/root/Configs/1Panel:/root/Configs/1Panel` 挂载，然后启动这个容器，执行`docker exec ID /app/reinstall.sh`

**Reply by purainity** [reply]
> 应该是IstoreOS安装时映射的问题，iStoreOS启动1Panel时设置的Docker映射里有`/root/Configs/1Panel:/iStorePane`，前面的目录是安装时配置的，后面的目录是固定的，到容器内的`/iStorePane`
> 
> > 见 https://github.com/linkease/openwrt-app-actions/blob/24f01cb47525e2e8c0677ef0ec8563a1c8487348/applications/luci-app-istorepanel/root/usr/libexec/istorec/istorepanel.sh#L55
> 
> 所以1Panel在内部创建的OpenResty目录`/root/Configs/1Panel/1panel/apps/openresty/openresty` 并没有映射到宿主机上(iStoreOS)
> 
> 而1Panel使用Docker创建OpenResty容器的时候，使用的是宿主机的目录，所以找不到文件
> 
> 
> 解决办法：
> 
> IstoreOS里面编辑1panel的那个容器，新增一个`/root/Configs/1Panel:/root/Configs/1Panel` 挂载，然后启动这个容器，执行`docker exec ID /app/reinstall.sh`
> 

试了下，好像还是不行

录屏：

https://github.com/user-attachments/assets/d35474ea-73df-454f-b399-9b2038a55024

**Comment by purainity** [comment]
> 应该是IstoreOS安装时映射的问题，iStoreOS启动1Panel时设置的Docker映射里有`/root/Configs/1Panel:/iStorePane`，前面的目录是安装时配置的，后面的目录是固定的，到容器内的`/iStorePane`
> 
> > 见 https://github.com/linkease/openwrt-app-actions/blob/24f01cb47525e2e8c0677ef0ec8563a1c8487348/applications/luci-app-istorepanel/root/usr/libexec/istorec/istorepanel.sh#L55
> 
> 所以1Panel在内部创建的OpenResty目录`/root/Configs/1Panel/1panel/apps/openresty/openresty` 并没有映射到宿主机上(iStoreOS)
> 
> 而1Panel使用Docker创建OpenResty容器的时候，使用的是宿主机的目录，所以找不到文件
> 
> 
> 解决办法：
> 
> IstoreOS里面编辑1panel的那个容器，新增一个`/root/Configs/1Panel:/root/Configs/1Panel` 挂载，然后启动这个容器，执行`docker exec ID /app/reinstall.sh`
> 

今天偶然发现 1panel 里显示的文件，但是在宿主机里就成了目录了，我怀疑就是这个的问题

![Screenshot_2024-11-07-19-58-20-695_com.android.browser.jpg](https://github.com/user-attachments/assets/2f4a4644-e90d-4a17-aed8-403854384e30)

![Screenshot_2024-11-07-20-01-10-434_com.android.browser.jpg](https://github.com/user-attachments/assets/31917ace-7aba-4bea-85c9-1552a5f3c130)

![Screenshot_2024-11-07-20-03-13-689_com.android.browser.jpg](https://github.com/user-attachments/assets/a9c05c00-2f0a-4394-abef-9a9677d583f7)

![Screenshot_2024-11-07-20-02-50-582_com.termux.jpg](https://github.com/user-attachments/assets/ed6d8040-24df-4993-8e5c-d08b8e5d5116)

![Screenshot_2024-11-07-20-01-03-469_com.mixplorer.silver.jpg](https://github.com/user-attachments/assets/d4ed8c14-2e3c-4411-a29b-ea6cdd4ae572)

**Comment by purainity** [comment]
> > 应该是IstoreOS安装时映射的问题，iStoreOS启动1Panel时设置的Docker映射里有`/root/Configs/1Panel:/iStorePane`，前面的目录是安装时配置的，后面的目录是固定的，到容器内的`/iStorePane`
> > 
> > > 见 https://github.com/linkease/openwrt-app-actions/blob/24f01cb47525e2e8c0677ef0ec8563a1c8487348/applications/luci-app-istorepanel/root/usr/libexec/istorec/istorepanel.sh#L55
> > 
> > 所以1Panel在内部创建的OpenResty目录`/root/Configs/1Panel/1panel/apps/openresty/openresty` 并没有映射到宿主机上(iStoreOS)
> > 
> > 而1Panel使用Docker创建OpenResty容器的时候，使用的是宿主机的目录，所以找不到文件
> > 
> > 
> > 解决办法：
> > 
> > IstoreOS里面编辑1panel的那个容器，新增一个`/root/Configs/1Panel:/root/Configs/1Panel` 挂载，然后启动这个容器，执行`docker exec ID /app/reinstall.sh`
> > 
> 
> 今天偶然发现 1panel 里显示的文件，但是在宿主机里就成了目录了，我怀疑就是这个的问题
> 
> ![Screenshot_2024-11-07-19-58-20-695_com.android.browser.jpg](https://github.com/user-attachments/assets/2f4a4644-e90d-4a17-aed8-403854384e30)
> 
> ![Screenshot_2024-11-07-20-01-10-434_com.android.browser.jpg](https://github.com/user-attachments/assets/31917ace-7aba-4bea-85c9-1552a5f3c130)
> 
> ![Screenshot_2024-11-07-20-03-13-689_com.android.browser.jpg](https://github.com/user-attachments/assets/a9c05c00-2f0a-4394-abef-9a9677d583f7)
> 
> ![Screenshot_2024-11-07-20-02-50-582_com.termux.jpg](https://github.com/user-attachments/assets/ed6d8040-24df-4993-8e5c-d08b8e5d5116)
> 
> ![Screenshot_2024-11-07-20-01-03-469_com.mixplorer.silver.jpg](https://github.com/user-attachments/assets/d4ed8c14-2e3c-4411-a29b-ea6cdd4ae572)
> 
> 

1Panel 给 OpenResty 创建的配置文件在 1Panel 里面看是正常的，但是在宿主机里全是文件夹。在宿主机里手动替换为正确的配置文件后 OpenResty 就能正确运行了。

但是目前 1Panel 看到的 OpenResty 配置文件和宿主机上真实存在的配置文件仍然不是同一个。 对1Panel 的网站设置进行编辑没有任何作用。而且 1Panel 里看到的配置文件到底存放在宿主机的哪里依旧未知。

**Comment by jannson** [comment]
我记得这个问题我改了，后来才看到的你的 ISSUE，应该好了吧？

**Reply by purainity** [reply]
我手动改了挂载路径正常了，但是如果直接安装的话还是会出现没挂载上的问题

**Comment by ming883** [comment]
设备绿联，系统是最新的pro，同样的问题。

**Reply by purainity** [reply, marked as answer]
已解决，停止原先的容器，删除宿主机内 `/ext_overlay/Configs/1Panel/1panel` 目录，新建一个容器添加下面的挂载路径，启动后执行容器内的 `/app/reinstall.sh` 就可以了

```
/sys/fs/cgroup:/sys/fs/cgroup
/var/run:/var2/run
/ext_overlay/Configs/1Panel:/ext_overlay/Configs/1Panel
/ext_overlay/Configs/1Panel:/iStorePanel
/mnt:/mnt:rslave
```

![Screenshot_2024-12-28-09-59-24-622_com yjllq chrome beta](https://github.com/user-attachments/assets/71422e16-f48a-48a9-a4ea-58821be51072)
